### PR TITLE
Inject pods into other deployments

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
To be able to colocate the jwtproxy with containers requiring secure access, we need to merge the jwtproxy pod into the deployments/pods of those containers. This PR tries to implement that.

This does not work yet!

It seems like UniqueNamesProvisioner is having issues replacing the to-be-mounted configmap names with their unique replacements. An unknown number of other problems have not yet been identified :) Also, tests are more probably royally broken - I haven't even tried to run them at this early stage of prototyping.

### What issues does this PR fix or reference?
#15651

### CODE REVIEW

**Comment by mshaposhnik** [comment]
can you add `PodMerger` class &

**Comment by metlos** [comment]
ci-build

**Comment by metlos** [comment]
ci-build

**Comment by metlos** [comment]
crw-ci-test

**Comment by metlos** [comment]
Note that the happy path failure is caused by https://github.com/eclipse/che/issues/16017 and so is independent of this PR.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java [hunk 1]
```diff
@@ -141,4 +154,33 @@ private Deployment createEmptyDeployment(String name) {
...
+  private String mergeServiceAccountName(@Nullable String a, @Nullable String b)
+      throws ValidationException {
+    return nonNullOrEqual(a, b, "Cannot merge pods with different service account names: %s, %s");
+  }
+
+  @Nullable
+  private static <T> T nonNullOrEqual(@Nullable T a, @Nullable T b, String errorMessageTemplate)
+      throws ValidationException {
+    if (a == null) {
+      return b;
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:177]: can potentially return `null` if both are. is that ok ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java [hunk 2]
```diff
@@ -356,6 +403,14 @@ public void setMetadata(ObjectMeta podMeta) {
...
     }
 
+    public PodRole getRole() {
+      return role;
+    }
+
+    public void setRole(PodRole role) {
+      this.role = role;
+    }
+
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:415]: may be helper methods like `isInjectable()` and `isDeployment()` cus i see a lot of `getRole() ==...` constructions

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 3]
```diff
@@ -676,25 +682,101 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
...
+        storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
+      } else {
+        try {
+          injectables.add(new PodData(toCreate));
+          Deployment merged = podMerger.merge(injectables);
+          merged.getMetadata().setName(toCreate.getMetadata().getName());
+
+          final Pod createdPod = namespace.deployments().deploy(merged);
+          LOG.debug("Creating pod '{}' in workspace '{}'", toCreateMeta.getName(), workspaceId);
+          storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: may we externalize that from condition (and LOG.debug) because i see the difference only in `createdPod` variable, so seems we can only compute it under `if` condition and `storeStartingMachine` can be common

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java (hunk 1 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:177]: It can be null, as suggested by the `@Nullable` on the method.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java (hunk 2 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:415]: Not sure that it would improve the readability if we had `pod.isDeployment()` instead of `pod.getRole() == PodRole.DEPLOYMENT`. IMHO the latter is actually more clear in the intention.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java (hunk 1 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:177]: ok method name starts with `notNull` so little confusing

**Comment by metlos** [comment]
ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java [hunk 4]
```diff
@@ -141,4 +154,33 @@ private Deployment createEmptyDeployment(String name) {
...
+    return nonNullOrEqual(a, b, "Cannot merge pods with different service accounts: %s, %s");
+  }
+
+  private String mergeServiceAccountName(@Nullable String a, @Nullable String b)
+      throws ValidationException {
+    return nonNullOrEqual(a, b, "Cannot merge pods with different service account names: %s, %s");
+  }
+
+  @Nullable
+  private static <T> T nonNullOrEqual(@Nullable T a, @Nullable T b, String errorMessageTemplate)
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:174]: any reason it is static?

Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:174]: I think this is more readable without nesting:
```
    if (a == null) {
      return b;
    } else if (b == null || a.equals(b)) {
      return a;
    } else {
      throw new ValidationException();
    }
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/CertificateProvisioner.java [hunk 5]
```diff
@@ -83,15 +84,17 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
...
-      Optional<Volume> certVolume =
-          pod.getSpec()
-              .getVolumes()
-              .stream()
-              .filter(v -> v.getName().equals(CHE_SELF_SIGNED_CERT_VOLUME))
-              .findAny();
-
-      if (!certVolume.isPresent()) {
-        pod.getSpec().getVolumes().add(buildCertSecretVolume(selfSignedCertSecretName));
+      if (pod.getRole() == PodRole.DEPLOYMENT) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/CertificateProvisioner.java:86]: I think this is equivalent, but shorter and simpler 
```
        if (pod.getSpec().getVolumes().stream()
            .noneMatch(v -> v.getName().equals(CHE_SELF_SIGNED_CERT_VOLUME))) {
          pod.getSpec().getVolumes().add(buildCertSecretVolume(selfSignedCertSecretName));
        }
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSslCertificateProvisioner.java [hunk 6]
```diff
@@ -103,15 +104,17 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
...
-      Optional<Volume> certVolume =
-          pod.getSpec()
-              .getVolumes()
-              .stream()
-              .filter(v -> v.getName().equals(CHE_GIT_SELF_SIGNED_VOLUME))
-              .findAny();
-
-      if (!certVolume.isPresent()) {
-        pod.getSpec().getVolumes().add(buildCertVolume(selfSignedCertConfigMapName));
+      if (pod.getRole() != PodRole.INJECTABLE) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSslCertificateProvisioner.java:106]: again can be simplified with `stream().noneMatch()`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java [hunk 7]
```diff
@@ -240,41 +269,37 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
...
-      ContainerPort containerPort;
-      if (exposedOpt.isPresent()) {
-        containerPort = exposedOpt.get();
-      } else {
-        containerPort =
-            new ContainerPortBuilder().withContainerPort(port).withProtocol(protocol).build();
-        container.getPorts().add(containerPort);
-      }
+  private void exposeInContainerIfNeeded(ServicePort servicePort) {
+    Optional<ContainerPort> exposedOpt =
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java]: `stream().noneMatch()`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposer.java [hunk 8]
```diff
@@ -26,12 +30,33 @@
...
+   *
+   * <p>The exposer may choose to not create such service or to create a service for only a subset
+   * of the ports.
+   *
+   * @param allSecurePorts the secure ports on the container
+   * @param pod the pod containing the container
+   * @return a service to put "in front of" the pod to service the ports or null if no such service
+   *     should be created.
+   */
+  @Nullable
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposer.java]: I prefer `Optional` over `@Nullable`. I think it's a good fit here

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java [hunk 9]
```diff
@@ -60,19 +64,35 @@ public JwtProxySecureServerExposer(
     this.proxyProvisioner = jwtProxyProvisionerFactory.create(identity);
   }
 
+  @Override
+  public Service createService(
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java]: please javadoc why this is just `return null`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java [hunk 10]
```diff
@@ -175,57 +172,89 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
...
-            .withName(generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + '-' + machineName)
-            .withMachineName(machineName)
-            .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
-            .withPorts(new ArrayList<>(servicePorts))
-            .withServers(internalServers)
-            .build();
+    Map<String, ServerConfig> allNonSecureServers = new HashMap<>(internalServers);
+    allNonSecureServers.putAll(externalServers);
+    // only explicitly expose the service for the non-secure servers
+    if (!servicePorts.isEmpty()) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java]: can we extract this block to separate method? It's very hard to follow such long method doing so many things

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java [hunk 11]
```diff
@@ -175,57 +172,89 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
...
+              matchedExternalServers,
+              (serverId, srvrs) -> {
+                externalServerExposer.expose(
+                    k8sEnv, machineName, serviceName, serverId, servicePort, srvrs);
+              });
+        }
+      }
+    }
+
+    if (!secureServers.isEmpty()) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java]: can we extract this block to separate method? It's very hard to follow such long method doing so many things

**Review by sparkoo — COMMENTED** [review]
Mostly some refactoring tips. Logic looks ok, but I didn't get deep into it. It's quite hard to follow.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 12]
```diff
@@ -676,25 +682,91 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
...
     final String workspaceId = getContext().getIdentity().getWorkspaceId();
     LOG.debug("Begin pods creation for workspace '{}'", workspaceId);
+    PodMerger podMerger = new PodMerger();
+    Map<String, Map<String, Pod>> injectablePods = environment.getInjectablePodsCopy();
     for (Pod toCreate : environment.getPodsCopy().values()) {
       ObjectMeta toCreateMeta = toCreate.getMetadata();
-      final Pod createdPod = namespace.deployments().deploy(toCreate);
+      List<PodData> injectables = getAllInjectablePods(toCreate, injectablePods);
+
+      Pod createdPod;
```
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:691]: Look again :) While the code is indeed very similar, one loop operates on pods and the other on deployments.

The only code that could be extracted from the two loops is the `try` block in the `else` which could have been a method that we would pass the pod data created from either the pod or deployment. Since these are the only 2 occurrences of the logic and they're both in the same method very close together, I didn't consider it necessary to extract a method for the 4 lines of logic.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java (hunk 4 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java:174]: It is private and doesn't use any instance variables.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 13]
```diff
@@ -676,25 +682,91 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
...
+    return getAllInjectablePods(
+        podToCreate.getMetadata(), podToCreate.getSpec().getContainers(), injectables);
+  }
+
+  private List<PodData> getAllInjectablePods(
+      ObjectMeta toCreateMeta,
+      List<Container> toCreateContainers,
+      Map<String, Map<String, Pod>> injectables) {
+    Set<String> machinesInPod = getMachineNames(toCreateMeta, toCreateContainers);
+    return machinesInPod
```
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: The code coverage says it is tested :) I tried hard to write a transformation that would do it with one collect but in the end I decided this way would be better. One way to do this with 1 collect would be to wrap all entries in a custom object that would reimplement `equals/hashCode` based on the pod name and do `distinct` on the stream prior to collecting to list.

So the cost of doing that would be to wrap all intermediate entries in objects part of which would then be thrown away and creating an intermediate set in the `distinct()` call.

The solution as it is at the moment does away with the need for wrapping the entries with a custom object and creates a map explicitly.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/CertificateProvisioner.java (hunk 5 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/CertificateProvisioner.java:86]: Yeah, I just wrapped existing code in an `if` in trying to keep the amount of change to the minimum. But you're right, this can be simplified.

**Review by sparkoo — APPROVED** [review]

**Review by sparkoo — APPROVED** [review]
magnificent, thanks

**Comment by sparkoo** [comment]
double approve, wow

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
oh c'mon, this is not fair! I got hit by a failure by timeout twice in a row.. :roll_eyes:

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
the only test failing is "delete running workspace" which definitely works when tried manually. Retrying the CI tests one more time, maybe this time it will be fast enough to not timeout on it.

**Comment by metlos** [comment]
ci-test
